spec/compiler/semantic/module_spec.cr

```diff
@@ -226,15 +226,183 @@ describe "Semantic: module" do
         end
       end
 
-      class Baz(X)
+      class Bar(U)
+        include Foo(self)
+      end
+
+      Bar(Int32).new.foo
+      ") { generic_class("Bar", int32).metaclass }
+  end
+
+  it "includes generic module with self, and inherits it" do
+    assert_type("
+      module Foo(T)
+        def foo
+          T
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      class Baz < Bar(Int32)
+      end
+
+      Baz.new.foo
+      ") { types["Baz"].metaclass }
+  end
+
+  it "includes generic module with self (check argument type, success)" do
+    assert_type("
+      module Foo(T)
+        def foo(x : T)
+          x
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      Bar(Int32).new.foo Bar(Int32).new
+      ") { generic_class("Bar", int32) }
+  end
+
+  it "includes generic module with self (check argument superclass type, success)" do
+    assert_type("
+      module Foo(T)
+        def foo(x : T)
+          x
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      class Baz < Bar(Int32)
+      end
+
+      Bar(Int32).new.foo Baz.new
+      ") { types["Baz"] }
+  end
+
+  it "includes generic module with self (check argument type, error)" do
+    assert_error "
+      module Foo(T)
+        def foo(x : T)
+          x
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      class Baz1 < Bar(Int32)
+      end
+
+      class Baz2 < Bar(Int32)
+      end
+
+      Baz1.new.foo Baz2.new
+      ", "no overload matches"
+  end
+
+  it "includes generic module with self (check argument superclass type, error)" do
+    assert_error "
+      module Foo(T)
+        def foo(x : T)
+          x
+        end
+      end
+
+      class Bar(U)
+        include Foo(self)
+      end
+
+      class Baz < Bar(Int32)
+      end
+
+      Baz.new.foo Bar(Int32).new
+      ", "no overload matches"
```

Comment:
Why should that error? I'd guess that `T` is `Bar(Int32) ` here, if not what is T?